File: Array.cpp
#include "header.h"
#include <iostream>

using namespace std;

// пустой массив
Array::Array(int initialCapacity) {
    head = nullptr;
    size = 0;
    capacity = initialCapacity; // максимально допустимое количество элементов
}

// Сложность: O(n)
void Array::push(int index, string value) {
    if (index < 0 || index > size || size >= capacity) { // проверяем корректность индекса и достаточность емкости
        cout << "Невозможно добавить элемент. Ошибка индекса или превышен максимальный размер." << endl;
        return;
    }

    Node* newNode = new Node(value); // создаем новый узел с переданным значением
    if (index == 0) { // если элемент вставляется в начало списка
        newNode->next = head; // следующий элемент нового узла указывает на текущую голову списка
        head = newNode; // голова обновляется на новый узел
    }
    else {
        Node* current = head; // начинаем с головы списка
        for (int i = 0; i < index - 1; i++) { // проходим до узла перед нужным индексом
            current = current->next; // переходим к следующему узлу
        }
        newNode->next = current->next; // новый узел указывает на следующий узел
        current->next = newNode; // предыдущий узел указывает на новый узел
    }
    size++;
}

// Сложность: O(n)
void Array::pushEnd(string value) {
    push(size, value);  // просто вызываем push, передавая размер списка как индекс для вставки в конец
}

// Сложность: O(n)
string Array::get(int index) {
    if (index < 0 || index >= size) { // проверка на корректность индекса
        return "Неверный индекс"; // возвращаем пустую строку
    }
    Node* current = head; // начинаем с головы
    for (int i = 0; i < index; i++) { // движемся к узлу с нужным индексом
        current = current->next;
    }
    return current->data; // возвращаем данные узла на нужной позиции
}

// Сложность: O(n)
void Array::pop(int index) {
    if (index < 0 || index >= size) {  // проверка на корректность индекса
        cout << "Неверный индекс!" << endl;
        return;
    }

    Node* toDelete = nullptr;  // указатель на удаляемый узел
    if (index == 0) {  // если удаляем первый элемент
        toDelete = head;  // запоминаем текущую голову
        head = head->next;  // обновляем голову
    }
    else {
        Node* current = head;  // начинаем с головы
        for (int i = 0; i < index - 1; i++) {  // движемся к узлу перед индексом
            current = current->next;
        }
        toDelete = current->next;  // узел для удаления
        current->next = toDelete->next;  // обходим удаляемый узел
    }
    delete toDelete;  // освобождаем память удаленного узла
    size--;  // уменьшаем размер списка
}

// Сложность: O(n)
void Array::replace(int index, string value) {
    if (index < 0 || index >= size) {  // проверка индекса
        cout << "Неверный индекс!" << endl;
        return;
    }

    Node* current = head;  // начинаем с головы

    for (int i = 0; i < index; i++) {  // движемся к нужному индексу
        current = current->next;
    }
    current->data = value;  // заменяем данные узла
}

// Сложность: O(1)
int Array::length() {
    return size;
}

// Сложность: O(n)
void Array::display() {
    if (head == nullptr) {
        cout << "Массив пуст!" << endl;
        return;
    }

    Node* current = head;  // начинаем с головы
    while (current) {  // пока есть элементы
        cout << current->data << " ";
        current = current->next;  // переходим к следующему узлу
    }
    cout << endl;
}

File: go_lr1/array.go
package main

import "fmt"

type Array struct {
	data     []string
	size     int
	capacity int
}

func NewArray() *Array {
	return &Array{data: make([]string, 10), size: 0, capacity: 10}
}

func (a *Array) PushEnd(value string) {
	if a.size == a.capacity {
		fmt.Println("Массив переполнен!")
		return
	}
	a.data[a.size] = value
	a.size++
}

func (a *Array) Push(index int, value string) {
	if index < 0 || index > a.size {
		fmt.Println("Индекс вне диапазона!")
		return
	}
	if a.size == a.capacity {
		fmt.Println("Массив переполнен!")
		return
	}
	for i := a.size; i > index; i-- {
		a.data[i] = a.data[i-1]
	}
	a.data[index] = value
	a.size++
}

func (a *Array) Get(index int) string {
	if index < 0 || index >= a.size {
		fmt.Println("Индекс вне диапазона!")
		return ""
	}
	return a.data[index]
}

func (a *Array) Pop(index int) {
	if index < 0 || index >= a.size {
		fmt.Println("Индекс вне диапазона!")
		return
	}
	for i := index; i < a.size-1; i++ {
		a.data[i] = a.data[i+1]
	}
	a.size--
}

func (a *Array) Replace(index int, value string) {
	if index < 0 || index >= a.size {
		fmt.Println("Индекс вне диапазона!")
		return
	}
	a.data[index] = value
}

func (a *Array) Length() int {
	return a.size
}

func (a *Array) Display() {
	if a.size == 0 {
		fmt.Println("Массив пуст!")
		return
	}
	for i := 0; i < a.size; i++ {
		fmt.Print(a.data[i], " ")
	}
	fmt.Println()
}
